transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/jobscheduling/WorkManagerScheduler.java

```
@@ -0,0 +1,114 @@
+import androidx.work.WorkRequest;
+import com.google.android.datatransport.runtime.TransportContext;
+import com.google.android.datatransport.runtime.logging.Logging;
+import com.google.android.datatransport.runtime.scheduling.persistence.EventStore;
+import com.google.android.datatransport.runtime.util.PriorityMapping;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+@RequiresApi(api = Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
+public class WorkManagerScheduler implements WorkScheduler {
+  private static final String LOG_TAG = "WorkManagerScheduler";
+
+  static final String ATTEMPT_NUMBER = "attemptNumber";
+  static final String BACKEND_NAME = "backendName";
+  static final String EVENT_PRIORITY = "priority";
+  static final String EXTRAS = "extras";
+  private final Context context;
+
+  private final EventStore eventStore;
+
+  private final SchedulerConfig config;
+
+  public WorkManagerScheduler(
+      Context applicationContext, EventStore eventStore, SchedulerConfig config) {
+    this.context = applicationContext;
+    this.eventStore = eventStore;
+    this.config = config;
+  }
+
+  @Override
+  public void schedule(TransportContext transportContext, int attemptNumber) {
+    schedule(transportContext, attemptNumber, false);
+  }
+
+  @Override
+  public void schedule(TransportContext transportContext, int attemptNumber, boolean force) {
+    WorkManager manager = WorkManager.getInstance(context);
+
+    int jobId = WorkScheduler.getJobId(context, transportContext);
+    if (!force) {
+      try {
+        for (WorkInfo info : manager.getWorkInfosByTag("transport-" + jobId).get()) {
+          if (!info.getState().isFinished()) {
+            Logging.d(
+                LOG_TAG,
+                "Upload for context %s is already scheduled. Returning...",
+                transportContext);
+            return;
+          }
+        }
+      } catch (Exception e) {
```

Could you add a comment here on why is OK to ignore any exception?

Also, is there a more specific Exception we could use ?

---

The WorkManager API shouldn't be throwing the handful of exceptions on the .get(), but if something changes, I've made it rethrow a runtime exception.

---

It has 3 possible exception (from the refdocs)

> Throws:
> CancellationException – if the computation was cancelled
> ExecutionException – if the computation threw an exception
> InterruptedException – if the current thread was interrupted while waiting

Are we OK crashing the app if any of these happen?